pywbemtools/pywbemcli/_cmd_connection.py

```diff
@@ -251,21 +272,25 @@ def connection_select(context, name):
                      default='all'))
 @add_options(verify_option)
 @click.pass_obj
```

Comment:
**Discussion point:**

As part of the discussion on --name on the save command, I stumbled across this question:

Why does "connection add" have command options for specifying the connection when the same options already exist as general options? I think "add" could use the general options for what it does. The only thing we need to watch out for is the current connection (that only exists in interactive mode), which we don't want to change, but I think when used in interactive mode, a "connection add" has its options in another context than the outermost context. We can try that out to make sure. If we can achieve that the current connection is not affected, then I suggest to change "add" to use its input connection from the general options.

Reply:
**DISCUSSION**

I agree that we two similar commands to get connection definitions into the connections file, 'add' and 'save'.  Originally, I did add so that it could modify an existing connection but then figured out that there were real issues here in that the general options do not really allow that.  For example, how could you modify a connection to set no_verify to False with the existing options( --no_verify sets it to True, but the only other options not to supply the option which means nothing more than option not defined). There is no way to separate option value is None vs option not supplied and a modify would depend on that difference to determine what variables to set.) 

We would have to modify many of the options in add so that the user had more detailed control over them ( ex. set --output-format to "None", set --no-verify to false) which would mean a real modify command not just an add that duplicates the general options.

So, why do we need the add at all?  You can use the general options in the interactive mode to define a server and just use 'connection save<name>'  to persist it with a name.

On startup, the general options defines what will be the current server and if the command 'save <name> saves that connection. 

In interactive mode, it could use the general options and 'save <name>'  command to create and save a connection but not make it current. Note that save DOES NOT do this today since we do not capture these general option inputs that define a connection in interactive mode.

But the add does create and persist a connection definition without affecting the current. The save saves the current.

The only command that  changes the current connection in interactive mode is the 'connection select' in that it sets the connection definition in the root click context.  NOTE: I just added a clarifying method to context_obj for that.  update_root_click_context

Reply:
Why is it that the "save" command cannot not process general options in interactive mode? It does process general options in command mode. Which is an inconsistency we should address anyway.

For example, today the use of general options on "save" in interactive mode is simply ignored, which is definitely unexpected for a user:

```
$ pywbemcli --server https://localhost
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli.
pywbemcli> -m tests/unit/simple_mock_model.mof connection save mock1
pywbemcli> connection list
WBEM server connections: (#: default, *: current)
+--------+-------------------+-------------+-----------+-------------+
| name   | server            | namespace   |   timeout | no-verify   |
|--------+-------------------+-------------+-----------+-------------|
| *mock1 | https://localhost | root/cimv2  |        30 | False       |
+--------+-------------------+-------------+-----------+-------------+
pywbemcli>          
```

While in command mode:
```
$ pywbemcli -m tests/unit/simple_mock_model.mof connection save mock1
$ pywbemcli connection list
WBEM server connections: (#: default, *: current)
+--------+-------------+-----------+-------------+----------------------------------+
| name   | namespace   |   timeout | no-verify   | mock-server                      |
|--------+-------------+-----------+-------------+----------------------------------|
| mock1  | root/cimv2  |        30 | False       | tests/unit/simple_mock_model.mof |
+--------+-------------+-----------+-------------+----------------------------------+
```
and of course it rejects using --server and -m at the same time.

Reply:
One more thought: If we allow that "save" can overwrite existing connection definitions, it (1) is what one expects from a save, and (2) it provides the modify functionality.

Reply:
Also, the fact that "save" sets the current connection is weird. There is also an issue with it. The following causes two current connections to be displayed:
```
$ pywbemcli                           
Enter 'help' for help, <CTRL-D> or ':q' to exit pywbemcli.
pywbemcli> connection add -m tests/unit/simple_assoc_mock_model.mof mock1
pywbemcli> connection add -m tests/unit/simple_mock_model.mof mock2
pywbemcli> connection list
WBEM server connections: (#: default, *: current)
+--------+-------------+-----------+-------------+----------------------------------------+
| name   | namespace   |   timeout | no-verify   | mock-server                            |
|--------+-------------+-----------+-------------+----------------------------------------|
| mock1  | root/cimv2  |        30 | False       | tests/unit/simple_assoc_mock_model.mof |
| mock2  | root/cimv2  |        30 | False       | tests/unit/simple_mock_model.mof       |
+--------+-------------+-----------+-------------+----------------------------------------+
pywbemcli> connection select mock1
"mock1" current
pywbemcli> connection list
WBEM server connections: (#: default, *: current)
+--------+-------------+-----------+-------------+----------------------------------------+
| name   | namespace   |   timeout | no-verify   | mock-server                            |
|--------+-------------+-----------+-------------+----------------------------------------|
| *mock1 | root/cimv2  |        30 | False       | tests/unit/simple_assoc_mock_model.mof |
| mock2  | root/cimv2  |        30 | False       | tests/unit/simple_mock_model.mof       |
+--------+-------------+-----------+-------------+----------------------------------------+
pywbemcli> connection save mock3
pywbemcli> connection list
WBEM server connections: (#: default, *: current)
+--------+-------------+-----------+-------------+----------------------------------------+
| name   | namespace   |   timeout | no-verify   | mock-server                            |
|--------+-------------+-----------+-------------+----------------------------------------|
| *mock1 | root/cimv2  |        30 | False       | tests/unit/simple_assoc_mock_model.mof |
| *mock3 | root/cimv2  |        30 | False       | tests/unit/simple_assoc_mock_model.mof |
| mock2  | root/cimv2  |        30 | False       | tests/unit/simple_mock_model.mof       |
+--------+-------------+-----------+-------------+----------------------------------------+
pywbemcli> 
```

Reply:
Not wierd, that is what happens if you don't copy before modifying.  
FIXED.  Also does not touch the current connection now.

Reply:
We agreed to try implementing the following behavior:
- in interactive mode, commands can get global options but they apply only to that one command and not to future commands.
- in interactive mode, global options specified on a command overrwrite the inherited global options from the command invocation. Conflicting options in the reslting effective set of global options are rejected.
- connection save allows saving to an already existing connection name, overwriting it without warning.
- we drop connection add.

Reply:
I propose to make the overwrite on save and the question of verify separate prs.

Reply:
These are all in separate PRs, as shown above.